### ISSUE TITLE
Which Promise libraries are supported?

### ISSUE THREAD

**Issue by hitode909** [graphql-perl/graphql-perl#32]
Do you have any supported Promise library other than `Mojo::Promise`?
I'm currently working on pure Plack application, so I want to use `GraphQL` without `Mojolicious`.

I tried some libraries, but they have compatibility problems.
- [Promises](https://metacpan.org/pod/Promises)
  - This library seems famous.
  - Its [Promise](https://metacpan.org/pod/Promises::Promise) object doesn't have `resolve` and `reject` method, so [is_Promise](https://github.com/graphql-perl/graphql-perl/blob/4dbf6064ce9d68400c222b20f935ea319c64d90e/lib/GraphQL/Type/Library.pm#L408-L415) returns false.
  - Before 51952c4f6e91b3d7e15e40d6f53e478fe55be854, `Promises` seems work fine with `GraphQL` because old version requires only `then` method.
- [Promise::ES6](https://metacpan.org/pod/Promise::ES6)
  - Its [resolve](https://metacpan.org/source/FELIPE/Promise-ES6-0.23/lib%2FPromise%2FES6%2FBackend%2FPP.pm#L60) receives only one scalar value(resolve($value)), not scalar values like [resolve(@values)](https://metacpan.org/release/Mojolicious/source/lib/Mojo/Promise.pm#L91).
  - GraphQL's [{promise_code}{all}](https://github.com/graphql-perl/graphql-perl/blob/4dbf6064ce9d68400c222b20f935ea319c64d90e/lib/GraphQL/Execution.pm#L405-L407) expects resolve(@values).

Which Promise libraries are supported? If you have some compatible Promise libraries, I'll try it.
Or, What do you think about changing the type definition of Promise to require only `then` method? I think this is not a bad idea because `GraphQL` doesn't call `resolve` or `reject` directly, It always calls via `promise_code`.
Thank you.

**Comment by mohawk2** [comment]
The `promise_code` was brought in to support `Future` (which also needed changing so its `then` would promote return values into `Future`s, as of its version 0.45). I think you're right that the `Promise` spec didn't keep up. I've loosened it and just released that as 0.45.

Please could you try again with those libraries? If you make `PromiseCode` that works with them, then if you either PR patches to the `PromiseCode` doc, or just put them on this issue so I can do it, then that would be amazing!

**Comment by hitode909** [comment]
@mohawk2
I tried [Promises](https://metacpan.org/pod/Promises) with GraphQL 0.45.
Then, It worked fine without any patches for the PromiseCode.

```perl
# use Promises like this
use Promises qw(collect resolved rejected);

# just call them from PromiseCode 
{
  all     => \&collect,
  resolve => \&resolved,
  reject  => \&rejected,
},
```

Thank you for your quick response! I'll close this issue.

**Comment by mohawk2** [comment]
Thanks! Added to the docs, released as 0.46.
